### PR TITLE
Make facets work with Solr 7

### PR DESCRIPTION
**Pull request by whikloj**
**JIRA Ticket**: 
 * https://jira.duraspace.org/browse/ISLANDORA-2201
 * https://jira.duraspace.org/browse/ISLANDORA-2271

# What does this Pull Request do?

We (U of Manitoba) want to upgrade the supporting software for Islandora, specifically Solr. We are trying with Solr 7 and found that date faceting has been removed. Instead we must use Range faceting only.

This PR adds in support for range facets. The framework was there so I just tried to fill in the bits and pieces.

# What's new?
By default I pass `facet.date` parameters also as `facet.range` as I could not see how to tell whether a facet was a date or range type in the query_processor. It did not seem to have a noticeable effect.

Inside I just adjusted the facet results, to account for the differences but re-use most of the code.

I am assuming we will be supporting all versions of Solr from 3 to 100 so this probably needs a lot of testing, but I am also not convinced most people will upgrade their Solr so maybe the primary focus is to ensure the current functionality is maintained.

# How should this be tested?

This is tough, I could not just point FedoraGSearch at Solr 7 as it now throws a stacktrace.
```
dk.defxws.fedoragsearch.server.errors.GenericSearchException: Mon Jul 09 20:17:08 UTC 2018 updateIndex NumberFormatException numFound=nseHeader
{  "responseHeader":{    "status":0,    "QTime":0,    "params":{      "q":"*:*",      "rows":"0"}},  "response":{"numFound":21,"start":0,"docs":[]  }}; nested exception is: 
	java.lang.NumberFormatException: For input string: "nseHeader"
	at dk.defxws.fgssolrremote.OperationsImpl.updateIndex(OperationsImpl.java:118)
	at dk.defxws.fedoragsearch.server.GenericOperationsImpl.updateIndex(GenericOperationsImpl.java:556)
	at dk.defxws.fedoragsearch.server.RESTImpl.updateIndex(RESTImpl.java:313)
	at dk.defxws.fedoragsearch.server.RESTImpl.doGet(RESTImpl.java:149)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:620)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.fcrepo.server.security.servletfilters.FilterSetup.doFilter(FilterSetup.java:247)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.fcrepo.server.security.servletfilters.FilterSetup.doFilter(FilterSetup.java:247)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.fcrepo.server.security.servletfilters.FilterSetup.doFilter(FilterSetup.java:247)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.fcrepo.server.security.servletfilters.FilterSetup.doFilter(FilterSetup.java:247)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
	at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:408)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1070)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:611)
	at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:314)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NumberFormatException: For input string: "nseHeader"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at java.lang.Integer.parseInt(Integer.java:615)
	at dk.defxws.fgssolrremote.OperationsImpl.updateIndex(OperationsImpl.java:116)
	... 37 more
```

 So I had to use my own indexer to get stuff into Solr, I didn't spend a lot of time debugging GSearch though so I might have missed a simple fix.

Anyways, my major concern is that this doesn't harm everyone else as I imagine the number of people upgrading their Solr is in the minority.

# Additional Notes:
Any additional information that you think would be helpful when reviewing this PR.

Example:
* Does this change require documentation to be updated? Probably, though from the user's side nothing has changed
* Does this change add any new dependencies? no
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? no
* Could this change impact execution of existing code? no

# Interested parties
@Islandora/7-x-1-x-committers @DiegoPino @giancarlobi

### CODE REVIEW

File: includes/query_processor.inc
Comment (DiegoPino) [inline includes/query_processor.inc]: Comment here needs an update

File: includes/query_processor.inc
Comment (DiegoPino) [inline includes/query_processor.inc]: Question here: seems like facet.range was introduced in Solr 3.3+, so is there really a need to maintain deprecated on solr 6.6 facet.date piece? can we just get rid of it?

File: includes/query_processor.inc
Comment (DiegoPino) [inline includes/query_processor.inc:243]: Feels like something is missing here? i see foreach happens on facet_array now but i don´t see any changed logic that supports that change further up.  I mean https://github.com/whikloj/islandora_solr_search/blob/5031f4ff93de89464cd3ec0e69831ee2a0dc9a04/includes/query_processor.inc#L201 is still there.

File: includes/results.inc
Comment (DiegoPino) [inline includes/results.inc:709]: Same as previous review, maybe we can just use ranges

File: includes/utilities.inc
Comment (DiegoPino) [inline includes/utilities.inc:637]: Since Solr 5.x there are also other changes in field types relevant to dates and ranges. One in particular for Multi valued dates i care for is
https://lucene.apache.org/solr/6_4_0/solr-core/org/apache/solr/schema/DateRangeField.html

**Review by DiegoPino — COMMENTED** [review]
Some comments for discussion

File: includes/query_processor.inc
Comment (whikloj) [reply includes/query_processor.inc]: I don't know, I just assume that someone somewhere is relying on a very old piece.

File: includes/query_processor.inc
Comment (whikloj) [reply includes/query_processor.inc:243]: See I added this back in [PR-300](https://github.com/Islandora/islandora_solr_search/pull/300) and I realized here that `$facet_dates` are already in `$facet_array` and this merge actually doesn't do anything.

File: includes/results.inc
Comment (whikloj) [reply includes/results.inc:709]: Same answer as above, but if people want me to try removing the date stuff I can certainly do that. Not sure if that should be part of this or separate though?

File: includes/query_processor.inc
Comment (DiegoPino) [reply includes/query_processor.inc]: Ok, cool we can discuss this on our next committers call. Won't block this pull but good to have consensus. Probably more needed to be really Solr 7 compliant in the future

File: includes/query_processor.inc
Comment (DiegoPino) [reply includes/query_processor.inc:243]: Yeah, makes sense, but looks like another ticket to me. Not saying it is not good to fix since you are here, but good to document when/why this change happened somewhere. So $facet_dates is empty? Not needed at all?

**Review by DiegoPino — COMMENTED** [review]
Jared. Looks good, concerns addressed except the thing with that array merge where i would love to see a ticket or some note here saying why it is being removed.  Will test with Solr 4.10 and 5.5 later tomorrow and then approve (if i approve too fast i'm pretty sure someone will just merge, has happened to me a lot this week and i really want to test with date data). Sorry for the delay but Solr is corner stone of islandora. Thanks a lot

File: includes/query_processor.inc
Comment (whikloj) [reply includes/query_processor.inc:243]: Updated with new ticket to resolve this https://jira.duraspace.org/browse/ISLANDORA-2271

File: includes/results.inc
Comment (adam-vessey) [inline includes/results.inc]: Comment is no longer relevant?

File: includes/results.inc
Comment (adam-vessey) [inline includes/results.inc]: I'm not entirely sure this is appropriate, as range fields can cover things that are not dates... integers or other numerical data... but here, we do some pretty explicit [date formatting](https://github.com/Islandora/islandora_solr_search/pull/342/files#diff-17a5945b1aa412dec3b39453ad29425fR940).

File: includes/query_processor.inc
Comment (adam-vessey) [inline includes/query_processor.inc]: Not sure these defaults are appropriate... do we have to do [some form of detection](https://github.com/Islandora/islandora_solr_search/pull/342/files#diff-2d69368f59ba81d5635cd8274dd66090R633), to determine if date-related parameters are appropriate? Also, brings up the question of what would be a reasonable default, otherwise, for numerical data and the like.

File: includes/query_processor.inc
Comment (adam-vessey) [inline includes/query_processor.inc]: Calculation of both where both are supported seems like an unnecessary strain on Solr.

File: includes/query_processor.inc
Comment (adam-vessey) [reply includes/query_processor.inc]: Almost... wondering if we should instead move to a factory, to facilitate the use of different implementations for different versions of Solr... the [version of Solr can be requested of Solr relatively easily](https://stackoverflow.com/a/48835496), hitting:
`http://whatever:port/solr/admin/info/system?wt=json`, giving something like (added in `indent=true` to the query parameters for formatting):
```json
{
  "responseHeader":{
    "status":0,
    "QTime":5},
  "mode":"std",
  "lucene":{
    "solr-spec-version":"4.10.4",
    "solr-impl-version":"4.10.4 1662817 - mike - 2015-02-27 16:43:07",
    "lucene-spec-version":"4.10.4",
    "lucene-impl-version":"4.10.4 1662817 - mike - 2015-02-27 16:38:43"},
[... a bunch of environment stuff ...]}
```

Would we then want to make `IslandoraSolrQueryProcessor` class just proxy out to whatever target implementation?

... Not sure... maybe going too far with abstraction fun, especially given this module supposedly just "supports" Solr 4.2+.

**Review by adam-vessey — COMMENTED** [review]
I'm... not sure mapping ranges directly as dates is quite appropriate... there could be other data types for which range faceting might be useful; otherwise, a couple of little comments.

**Comment by whikloj** [comment]
@adam-vessey you bring up some good points. I was not considering using a numerical field as a range, so that would cause a problem. I think I can handle that by checking the field type to see if it is a date type before applying the date formatting. As for the idea of a factory, I do like that idea in general I'm just not sure if 1) I'm the best person to try and implement it and 2) how much deviation there is between the various versions of Solr (especially the subset that Islandorians use).

File: includes/results.inc
Comment (whikloj) [reply includes/results.inc]: I'm not disagreeing with what you are saying here. Just that some range formatting seems better than none. Once we have someone using this with Solr 6+ it will be easier to see the holes. One suggestion would be to make a field a `facet_date` if its a `facet_range` but on a `islandora_solr_is_date_field()` field.

**Comment by whikloj** [comment]
I've expanded on this a bit and hopefully cleaned up some concerns.

I have reverted using the `processFacetRanges()` and left it empty and instead push `facet_range` types that are also date fields straight into the `processFacetDates()` function. So when someone wants to make an integer facet range, they can write the code for that. 

Alternatively if this seems too trouble some there is a much larger and definitely tougher PR (#345) which I created. I am trying to write a Solarium backend for it, but due to the way we let people mess around inside the IslandoraSolrQueryProcessor it is problematic.

File: includes/query_processor.inc
Comment (adam-vessey) [inline includes/query_processor.inc]: Should possibly be `protected` instead of `private`... facilitate extensibility?

File: includes/query_processor.inc
Comment (adam-vessey) [inline includes/query_processor.inc]: Same deal, `private` vs `protected`.

File: includes/results.inc
Comment (adam-vessey) [inline includes/results.inc]: Same deal, `private` vs `protected`.

**Review by adam-vessey — COMMENTED** [review]
Looks good... might've went and [did a thing](https://github.com/discoverygarden/gsearch/compare/master...adam-vessey:newer_compat) ([more specifically, one commit](https://github.com/discoverygarden/gsearch/commit/6e88084c88bd7a66ffbe5e465343f70d8142e771); would want to firm up some other things before going further with it), but... appeared to do the trick... Both Solr 4.10.4 and 7.5...

Could easily live with `private` for these... (could always widen the scope later, should it be needed)... What do we think?

**Comment by whikloj** [comment]
@adam-vessey the specific commit does not seem to be accessible by me. Though by the first link that appears to be around getting GSearch to work with newer versions of Solr, yes?

**Comment by whikloj** [comment]
As for `protected` versus `private`, I have no problem with going to `protected`. If we want to go to `public` then I'd want to add a getter/setter to make it easier if we need to alter the insides in future.

**Comment by adam-vessey** [comment]
@whikloj: Whoops on the commit... accidentally added a `0` in making the link (instead of the closing parenthesis... then added the closing parenthesis, I'm assuming?) Fixed, but also here: https://github.com/discoverygarden/gsearch/commit/6e88084c88bd7a66ffbe5e465343f70d8142e771

Yeah, that's pretty much it... newer Java (for the Base64 stuff), newer Solr, and a bit of an improvement to allow for relative HREFs in XSLTs (which still needs to be put through its paces).

**Review by adam-vessey — APPROVED** [review]

**Comment by bondjimbond** [comment]
@adam-vessey You approved this last week, still unmerged... is it good to go?

**Pull request merged**
